contracts/adapters/ibc/secret-ibc-hooks/src/contract.rs

```diff
@@ -0,0 +1,497 @@
+            // Remove the AckID <> in progress ibc transfer from storage
+            // and return immediately if the ibc transfer was successful
+            // since no further action is needed.
+            if success {
+                let ack_id: AckID = (&channel, sequence);
+                ACK_ID_TO_RECOVER_ADDRESS.remove(deps.storage, ack_id);
+
+                return Ok(Response::new().add_attribute("action", SudoType::Response));
+            }
+
+            (channel, sequence, SudoType::Error)
+        }
+        SudoMsg::IbcLifecycleComplete(IbcLifecycleComplete::IbcTimeout { channel, sequence }) => {
+            (channel, sequence, SudoType::Timeout)
+        }
+    };
+
+    // Get and remove the AckID <> in progress recover address from storage
+    let ack_id: AckID = (&channel, sequence);
+    let to_address = ACK_ID_TO_RECOVER_ADDRESS.load(deps.storage, ack_id)?;
+    ACK_ID_TO_RECOVER_ADDRESS.remove(deps.storage, ack_id);
+
+    // Get all coins from contract's balance, which will be the the
+    // failed ibc transfer coin and any leftover dust on the contract
+    let amount = deps.querier.query_all_balances(env.contract.address)?;
+
+    // If amount is empty, return a no funds to refund error
+    if amount.is_empty() {
+        return Err(ContractError::NoFundsToRefund);
+    }
+
+    // Create bank send message to send funds back to user's recover address
+    let bank_send_msg = BankMsg::Send { to_address, amount };
+
+    Ok(Response::new()
+        .add_message(bank_send_msg)
+        .add_attribute("action", sudo_type))
+}
+
+////////////////////////
+/// HELPER FUNCTIONS ///
+////////////////////////
+
+// Verifies the given memo is empty or valid json, and then adds the necessary
+// key/value pair to trigger the ibc hooks callback logic.
+fn verify_and_create_memo(memo: String, contract_address: String) -> ContractResult<String> {
+    // If the memo given is empty, then set it to "{}" to avoid json parsing errors. Then,
+    // get Value object from json string, erroring if the memo was not null while not being valid json
+    let mut memo: Value = serde_json_wasm::from_str(if memo.is_empty() { "{}" } else { &memo })?;
+
+    // Transform the Value object into a Value map representation of the json string
+    // and insert the necessary key value pair into the memo map to trigger
+    // the ibc hooks callback logic. That key value pair is:
+    // { "ibc_callback": <CALLBACK_CONTRACT_ADDRESS> }
+    //
+    // If the "ibc_callback" key was already set, this will override
+    // the value with the current contract address.
+    if let Value::Map(ref mut memo) = memo {
+        memo.insert(
+            Value::String("ibc_callback".to_string()),
+            Value::String(contract_address),
+        )
+    } else {
+        unreachable!()
+    };
+
+    // Transform the memo Value map back into a json string
+    let memo = serde_json_wasm::to_string(&memo)?;
+
+    Ok(memo)
+}
+
+fn register_tokens(
```

Comment:
q for my understanding: Can you link me to documentation or give me a tldr on what registering a token does?